python/cuml/common/base.pyx

```diff
@@ -0,0 +1,92 @@
+#
+# Copyright (c) 2019, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# cython: profile=False
+# distutils: language = c++
+# cython: embedsignature = True
+# cython: language_level = 3
+
+
+import cuml.common.handle
+import cuml.common.cuda
+
+
+class Base:
+    """
+    Base class for all the ML algos. It handles some of the common operations
+    across all algos. Every ML algo class exposed at cython level must inherit
+    from this class.
+    """
+
+    def __init__(self, handle=None, verbose=False):
+        """
+        Constructor. All children must call init method of this base class!
+
+        Parameters
+        ----------
+        handle : cuml.common.handle.Handle
+               If it is None, a new one is created just for this class
+        verbose : bool
+                Whether to print debug spews
+        """
+        if handle is None:
+            self.handle = cuml.common.handle.Handle()
+            self.stream = cuml.common.cuda.Stream()
+            self.handle.setStream(self.stream)
+        else:
+            self.handle = handle
+        self.verbose = verbose
+
+
+    def get_param_names(self):
+        """
+        Returns a list of parameter names owned by this class. It is expected
+        that every child class overrides this method and appends its extra set
+        of parameter that it needs. This will simplify the `get_params` and
+        `set_params` methods. (Refer below)
+        """
+        return ["handle", "verbose"]
+
+
+    def get_params(self, deep=True):
```

Comment:
I was thinking, it might be worth it to add the option of excluding certain parameters from the ones you get (perhaps with an optional parameter `exclude=[]`), so that users only interested in the model parameters don't get things like the `handle` and the `verbose`. 

Note: my above comment is regarding our compatibility with Dask-ML for hyper parameter optimization, so we'll be testing that the changes we introduce don't break our compatibility. Another way of assuring compatibility might be to only return hyperparameters of the model by default, and have a flag that would allow users interested in getting things like the handle additionally. But will expand once we get the chance to test.